### DISCUSSION TITLE
What does your backup stack look like?

### DISCUSSION THREAD

**Discussion by Guusggg** [inventree/InvenTree#2514] [General]
Hi everybody,

Lately I've been bothered by the fear of our company burning down and taking all the data we collected in InvenTree so far with it. This raised a question to me (as I'm basically a hobbyist, helping someone start up), what do you guys use for reliable backups of InvenTree?

Myself, I've been thinking about making cloud storage available to friends and family, and I would store the exports of InvenTree along with an SQL dump to be able to restore it. Do I need to think of more things? I think the general rule of data storage is; the more places, the better. 

Also do you guys automate InvenTree backups?

I really like to hear how you guys manage this!

Kind regards,

**Comment by eeintech** [comment]
At my company we have multiple bare metal installation of InvenTree, they are all backed up more or less in the same way:
* PostgresSQL dump => single TGZ file
* JSON dump (redundant to DB dump) using `invoke export-records` => single JSON file
* Archive media folder (for all the images, labels, etc.) => single TGZ file

Those 3 files are then backed up daily on our Git server, which is a separate server from where all instances are running (yes more places the better).
The VM running on this server is also backed-up weekly on a different location, which gives another backup option in case our Git server **and** VM fail at the same time.

Side note, I have tried restoring from both PostgresSQL dump and JSON dump and got the same result so it's probably not necessary to do both.

**Reply by matmair** [reply]
@eeintech I did a few test recoveries, JSON or Postgres is fine, you lose some admin-logs and stuff with postgres but nothing that is needed for InvenTree itself. JSON is good for the backup software I use so I chose that. Both seem redundant but that is kind of the point 😂

**Comment by matmair** [comment]
I have a cron job running that dumps the data as json and pushes it into a backup endpoint for veeam, the media path is set directly to a versioned fileshare. Also, the whole VM is replicated multiple times a day. In most orgs I deploy to they have systems on other floors / other buildings and cities as backup targets. But that gets managed by their IT, my personal InvenTree is only replicated to one DR.

**Comment by wolffam** [comment]
What's the suggested way to take a backup from a bare metal installation and ingest that backup into a newly provisioned Docker deployment of Inventree?

**Reply by SchrodingersGat** [reply]
@wolffam as long as it is running the same database, it should be pretty straight forward.

0. Ensure you have a backup of database and media files (see comments by @eeintech above)
1. Setup a new docker deployment - ensure that the "data" directory is mounted to your 'local' computer
2. Bring the docker containers down
3. Replace the newly created database and media files with your old backup ones
4. Bring the containers back online

That *should* be enough to get you up and going. Obviously make sure you have copies of your data backups.

**Reply by wolffam** [reply]
Thanks for the step by step @SchrodingersGat !  I wasn't sure if I needed to do any kind of database migration after the fact.  Sounds like there's no need for that if it's a drop in replacement

**Reply by SchrodingersGat** [reply]
If you are migrating the *database* you should not need to run migrations. If you are exporting and then re-importing data, you will.

**Reply by wolffam** [reply]
I've been fiddling with this a bit and am curious if there's an easy way to go about this.
I currently have a sqlite database on a bare metal instance.  I want to move all of that data into a "production" instance set up with the docker deployment.  I can get the latter instance operational and have the postgres container up and running but I am hitting a wall trying to move the sqlite database into the postgres container and ingest that data into the postgres DB.  Any tips?

**Reply by eeintech** [reply]
> I am hitting a wall trying to move the sqlite database into the postgres container and ingest that data into the postgres DB. Any tips?

Can you give more details/context on your migration attempt and issues you are encountering?

**Reply by matmair** [reply]
@wolffam I can imagine what you are running into. Here is what I would do:
1. Verify the new instance works
2. Export all data from the old instance via invoke export-records (you get a data.json from that)
3. Zip all media files on the old instance
4. Move data to location accessible by new instances invoke
5. Drop all data from new instance with invoke
6. Load dataset via invoke import-records
7. Copy zipped media files to media destination on new instance
8. Confirm everything works in new instance
9. Shut off old instance

**Reply by wolffam** [reply]
Sure!  In the docker production deployment, I have the postgres container up and running and can establish a shell inside the database as guided in the docker production docs.  I have a `json` file containing the data from my sqlite database that's located on the host machine.  I'm not sure how to get that data.json into the postgres container (I can probably figure that out through the docker-compose) but even at that point I'm not sure how I would ingest that data into the postgres database.

**Reply by matmair** [reply]
@wolffam running invoke import-records should palce everything in postgres. Have you tried that?

**Reply by wolffam** [reply]
When running `invoke migrate` I am getting `Is the server running on that host and accepting TCP/IP connections?` which makes sense because invoke would be looking for the server locally (but it's actually running in the docker container).

**Reply by matmair** [reply]
@wolffam where in the steps outlined by me are you?

**Reply by wolffam** [reply]
I think we got it figured out.  We were on step 6 pretty much.  We had to remap some of the defaults in the `config.yaml` so that it was correctly pointing at the database container and it seems to be configured correctly now. 

Thanks for thinking through it!

**Reply by wolffam** [reply]
The one remaining item I can't seem to figure out is whether or not there is a way to transfer user accounts from one deployment to another.  Does anyone know?  It's not a big deal to remake them but I figure I would ask first

**Reply by SchrodingersGat** [reply]
The user account data is stored in the same database, so it will be transferred across in the same process.

**Reply by matmair** [reply]
@wolffam if the migration process keeps presenting problems and you have time I would appreciate a short issue in the docs repo with what info is missing. This could become a common problem / question and I think we should address this in the docs.

**Reply by wolffam** [reply]
I think most of my issues were due to a lack of understanding of the docker process/compose in general.  Once I got a handle on that, the current docs were helpful enough to guide me the rest of the way
